src/webgpu/util/conversion.ts

```diff
@@ -248,6 +255,62 @@ export function packRGB9E5UFloat(r: number, g: number, b: number): number {
   return rMantissa | (gMantissa << 9) | (bMantissa << 18) | (biasedExp << 27);
 }
 
+/**
+ * Quantizes two f32s to f16 and then packs them in a u32
+ *
+ * This should implement the same behaviour as the builtin `pack2x16float` from
+ * WGSL.
+ *
+ * Caller is responsible to ensuring inputs are f32s
+ *
+ * @param x first f32 to be packed
+ * @param y second f32 to be packed
+ * @returns an array of possible results for pack2x16float. Elements are either
+ *          a number or undefined.
+ *          undefined indicates that any value is valid, since the input went
+ *          out of bounds.
+ */
+export function pack2x16float(x: number, y: number): (number | undefined)[] {
+  if (!isFiniteF16(x) || !isFiniteF16(y)) {
+    // This indicates any value is valid, so it isn't worth bothering
+    // calculating the more restrictive possibilities.
+    return [undefined];
+  }
+
+  // Generates all possible valid f16s from a given f32. Assumes FTZ for both the f32 and f16 value is allowed.
+  const generate_f16s = (n: number): number[] => {
+    const n_f32s = [n];
+    if (n !== 0 && isSubnormalNumberF32(n)) {
+      n_f32s.push(0);
+    }
+
+    const n_f16s = n_f32s.flatMap(correctlyRoundedF16);
+    if (!n_f16s.includes(0) && n_f16s.some(isSubnormalNumberF16)) {
+      n_f16s.push(0);
+    }
+
+    return n_f16s;
+  };
+
+  const x_f16s = generate_f16s(x);
+  const y_f16s = generate_f16s(y);
+
+  const f16_pairs = cartesianProduct(x_f16s, y_f16s);
+  const results = new Array<number>();
+  f16_pairs.forEach(p => {
+    assert(p.length === 2, 'cartesianProduct of 2 arrays returned an entry with not 2 elements');
+
+    const buf = new ArrayBuffer(4);
+    const buf_f16 = new Float16Array(buf);
+    const buf_u32 = new Uint32Array(buf);
```

Comment:
I'm unsure how hot this code is but ArrayBuffer/view allocations are notoriously expensive, so we should consider optimizing this. You can optimize it by just reusing these 3 objects for every call instead of making 3 new ones each time. We can wrap that into a small helper function that takes 2 f16s and returns one u32 and implements that optimization internally.